src/common/marlin_server.cpp

```diff
@@ -1299,3 +1299,57 @@ void host_prompt_do(const PromptReason type, const char *const pstr, const char
         break;
     }
 }
+
+/*****************************************************************************/
+//FSM_notifier
+FSM_notifier::data FSM_notifier::s_data;
+
+FSM_notifier::FSM_notifier(ClinetFSM type, uint8_t phase, cvariant8 min, cvariant8 max,
+    uint8_t progress_min, uint8_t progress_max, uint8_t var_id)
+    : temp_data(s_data) {
+    s_data.type = type;
+    s_data.phase = phase;
+    s_data.scale = static_cast<float>(progress_max - progress_min) / static_cast<float>(max - min);
+    s_data.offset = -static_cast<float>(min) * s_data.scale + static_cast<float>(progress_min);
+    s_data.progress_min = progress_min;
+    s_data.progress_max = progress_max;
+    s_data.var_id = var_id;
+    s_data.last_progress_sent = -1;
+}
+
+//static method
+//x = (actual - s_data.min) * s_data.scale + s_data.progress_min;
+//x = actual * s_data.scale - s_data.min * s_data.scale + s_data.progress_min;
+// s_data.offset == -s_data.min * s_data.scale + s_data.progress_min
+//x = actual * s_data.scale + s_data.offset;
+void FSM_notifier::SendNotification() {
+    if (s_data.type == ClinetFSM::_none)
+        return;
+
+    cvariant8 temp;
+    temp.attach(marlin_vars_get_var(&(marlin_server.vars), s_data.var_id));
+
+    float actual = static_cast<float>(temp);
+    actual = actual * s_data.scale + s_data.offset;
+
+    int progress = int(actual); //int - must be signed
+    if (progress < s_data.progress_min)
+        progress = s_data.progress_min;
+    if (progress > s_data.progress_max)
+        progress = s_data.progress_max;
+
+    // after first sent, progress can only rise
+    if ((s_data.last_progress_sent == uint8_t(-1)) || (progress > s_data.last_progress_sent)) {
+        s_data.last_progress_sent = progress;
+        fsm_change(s_data.type, s_data.phase, progress, 0);
+    }
+}
+
+FSM_notifier::~FSM_notifier() {
+    s_data = temp_data;
+}
+
+/*****************************************************************************/
+//ClientResponseHandler
+//define static member
+uint32_t ClientResponseHandler::server_side_encoded_response = -1;
```

Comment:
Please document meaning of negative value assigned to unsigned type.

Reply:
done